### PR TITLE
New PTE solver, totally refactored all the PTE code, new root find for inverse EOS calls

### PR DESCRIPTION
**Pull request by jdolence**
## PR Checklist

- [ ] Adds a test for any bugs fixed. Adds tests for new features.
- [ ] Format your changes by using the `make format` command after configuring with `cmake`.
- [ ] Document any new features, update documentation for changes made.
- [ ] Make sure the copyright notice on any files you modified is up to date.

### CODE REVIEW

**Comment by jdolence** [comment]
@dholladay00 @jonahm @jhp-lanl you all might be interested in this.

**Comment by jhp-lanl** [comment]
I'm still going through this, but I really like the cleanups so far! I was working on something similar a while back but I had to set it aside. I'll keep going through this, but so far it looks really good!

File: singularity-eos/closure/mixed_cell_models.hpp
Comment (jhp-lanl) [reply singularity-eos/closure/mixed_cell_models.hpp:214]: As a follow-up, this appears to have units of temperature whereas it appears to be used as if it is unitless.

**Review by jhp-lanl — CHANGES_REQUESTED** [review]
Overall looks good. Just a few questions for the most part

File: singularity-eos/closure/mixed_cell_models.hpp
Comment (jdolence) [reply singularity-eos/closure/mixed_cell_models.hpp:145]: I didn't actually add this.  Not sure who did or what it's used for, that's why I didn't delete it.  Should perhaps get to the bottom of this thought.

File: singularity-eos/closure/mixed_cell_models.hpp
Comment (jdolence) [reply singularity-eos/closure/mixed_cell_models.hpp:167]: I had imagined there might be other solvers with needs beyond normalizing the volume fractions.  Actually, for the `PTESolverRhoU` solver, we might want to be normalizing the energies as well.  Not in there yet, but I was also thinking of adding another solver that could act as a fallback if PTE failed.  Could, for example, try to just get pressure equilibrium and toss out temperature equilibrium.  One formulation I'm thinking of wouldn't even update volume fractions and so wouldn't need to normalize them.  So, in short, I call it `Fixup` because it's supposed to be a hook to apply solver specific special sauce that enforces certain things that would vary from solver to solver.  Make sense?

File: singularity-eos/closure/mixed_cell_models.hpp
Comment (jdolence) [reply singularity-eos/closure/mixed_cell_models.hpp:231]: This bit of code is one of the important improvements to the solver robustness.  It deserves a longer conversation than we should probably have in a PR discussion.  Short version is that this code is intended to provide an initial guess for volume fraction for each material such that it's on the high density side of the minimum of the pressure curve at the given temperature.  Let's discuss offline.  @chadmeyer @Yurlungur and I have already discussed.

File: singularity-eos/closure/mixed_cell_models.hpp
Comment (jdolence) [reply singularity-eos/closure/mixed_cell_models.hpp]: I don't think so.  I think this is the right definition of the error norm to be used with our line search.  Numerical Recipes actually has a pretty straightforward discussion of this stuff, if I recall.  Could be wrong though, so argue back if you think that's wrong.

File: singularity-eos/closure/mixed_cell_models.hpp
Comment (jdolence) [reply singularity-eos/closure/mixed_cell_models.hpp:213]: Yeah, the input temperatures need to not be nonsense.  zero is fine for some of the materials, like if they didn't exist in a cell before a particular pte call, but they shouldn't just be random garbage.  we could make a different choice.  for example, we could just always guess 300K or something to start, but my preference is just to force callers to have at least initialized everything with something sane.  The benefit of this is that hosts can provide excellent initial guesses with little effort (like the cell's previous temperature) which can have a large impact on PTE solver performance.  same can be said of density/volume fractions.  In our block prototype code, I'm seeing much less than one iteration of the PTE solver on average for simple problems because the initial guesses are excellent.

Comment (jdolence) [reply singularity-eos/closure/mixed_cell_models.hpp:213]: > As a follow-up, this appears to have units of temperature whereas it appears to be used as if it is unitless.

I'm not sure what you're referring to here.

File: singularity-eos/closure/mixed_cell_models.hpp
Comment (jdolence) [reply singularity-eos/closure/mixed_cell_models.hpp:311]: As written the `ResidualNorm` function always computes the norm of the `residual` member variable, so wouldn't immediately apply here.  It would be straightforward to have that function take a `Real *` and then it could work for this too.  On the other hand, this is like one line of code anyway.  Could go either way on this one.

File: singularity-eos/closure/mixed_cell_models.hpp
Comment (jhp-lanl) [reply singularity-eos/closure/mixed_cell_models.hpp:145]: Ah okay, I just saw green and assumed it was added. @jonahm-LANL I use this in the vector EOS lookups. Where do you think it should live?

File: singularity-eos/closure/mixed_cell_models.hpp
Comment (jhp-lanl) [reply singularity-eos/closure/mixed_cell_models.hpp:167]: Ah gotcha. That makes sense. Could you add a summary of that in a comment?

File: singularity-eos/closure/mixed_cell_models.hpp
Comment (jhp-lanl) [reply singularity-eos/closure/mixed_cell_models.hpp:213]: > > As a follow-up, this appears to have units of temperature whereas it appears to be used as if it is unitless.
> 
> I'm not sure what you're referring to here.

There are several places in the code like this
```c++
temp[m] *= Tnorm;
```
or
```c++
eos[m].RhoPmin(Tnorm * Tideal)
```
or
```c++
sie[m] = eos[m].InternalEnergyFromDensityTemperature(rho[m], Tnorm * Tideal, Cache[m]);
```
or
```c++
Real e_pert = eos[m].InternalEnergyFromDensityTemperature(rho_pert, Tnorm * Tequil, Cache[m]);
```
These expressions in my mind imply that either `Tnorm` or what it multiplies should be unitless. Looking at the code more carefully, it looks like `Tideal` may be unitless?
```c++
  PORTABLE_INLINE_FUNCTION
  void GetIdealPTE(Real &Pideal, Real &Tideal) const {
    Real rhoBsum = 0.0;
    Real Asum = 0.0;
    for (int m = 0; m < nmat; m++) {
      Asum += vfrac[m] * press[m] / temp[m];
      rhoBsum += rho[m] * vfrac[m] * sie[m] / temp[m];
    }
    Asum *= u_total/Tnorm;
    rhoBsum /= Tnorm;
    Tideal = u_total / rhoBsum / Tnorm;
    Pideal = Tideal * Asum / vfrac_total / u_total;
  }
```
  
But `Tequil` certainly looks to have units, right? I see it initialized here with units for example:
```c++
PORTABLE_INLINE_FUNCTION
Real Init() {
  InitBase();
  Residual();
  TryIdealPTE(this);
  Tequil = temp[0];
  return ResidualNorm();
}
```

File: singularity-eos/closure/mixed_cell_models.hpp
Comment (jdolence) [reply singularity-eos/closure/mixed_cell_models.hpp:167]: yeah, will do.

File: singularity-eos/closure/mixed_cell_models.hpp
Comment (jhp-lanl) [reply singularity-eos/closure/mixed_cell_models.hpp:213]: Okay so I think I understand my confusion a bit more. `temp[:]` is normalized basically everywhere, hence multiplying it by `Tnorm` in the `Finalize()` member function. However, I would assume that somewhere there should be a
```c++
for (int m = 0; m < nmat; m++) {temp[m] /= Tnorm;}
```
somewhere in the initialization. Am I missing it?

File: singularity-eos/closure/mixed_cell_models.hpp
Comment (jdolence) [reply singularity-eos/closure/mixed_cell_models.hpp:82]: Yeah this was leftover from mid-refactor.  Removed it.  Good catch.

**Comment by dholladay00** [comment]
do we still need test_pte_flag?

**Comment by jdolence** [comment]
@jhp-lanl @dholladay00 just need to confirm RIOT works with this on GPU and then I'll remove the `Draft` and I'd like to get this merged in ASAP.  Anyone care to approve?

**Comment by dholladay00** [comment]
It is also running on gitlab CI, it needs to pass those tests as well, which will also test GPU compatibility @jdolence

**Comment by jdolence** [comment]
> It is also running on gitlab CI, it needs to pass those tests as well, which will also test GPU compatibility @jdolence

The gitlab CI still doesn't work for my pushes because of some permissions thing that never got resolved.

File: singularity-eos/closure/mixed_cell_models.hpp
Comment (Yurlungur) [reply singularity-eos/closure/mixed_cell_models.hpp:139]: Somewhere in `singularity-eos/eos_base.hpp` or `singularity-eos/base/*.hpp` I think. Not sure it has a good home but I agree there should only be one of these.

File: singularity-eos/closure/mixed_cell_models.hpp
Comment (Yurlungur) [reply singularity-eos/closure/mixed_cell_models.hpp:164]: Yeah I saw what you were doing here @jdolence and didn't love the name `Fixup`, but I couldn't think of a better one.

File: singularity-eos/closure/mixed_cell_models.hpp
Comment (jhp-lanl) [reply singularity-eos/closure/mixed_cell_models.hpp:139]: I'm going to resolve for now. We can figure this out later

File: singularity-eos/closure/mixed_cell_models.hpp
Comment (Yurlungur) [reply singularity-eos/closure/mixed_cell_models.hpp:217]: Can we add some (debug only) checks that enforce that the input temperatures are not nonsense? I.e., check for negative, NaN or infinite inputs?

File: singularity-eos/closure/mixed_cell_models.hpp
Comment (Yurlungur) [inline singularity-eos/closure/mixed_cell_models.hpp:262]: :+1:

**Review by jhp-lanl — APPROVED** [review]
I want to see the results of the CI on gitlab, but otherwise I approve the changes.

File: utils/root-finding-1d/root_finding.hpp
Comment (Yurlungur) [inline utils/root-finding-1d/root_finding.hpp:150]: :+1:

**Review by Yurlungur — APPROVED** [review]
Overall I like this. The object-oriented design saves a lot of pain from repeated code, and I think the new solver looks very promising.

Additionally I like the additional guard rails for tension in PTE.

Some nitpicky comments below. Also we should double check that Regula Falsi is better than the old 1D root finder. (I hope it is. Great to have Regula Falsi in here!)

I triggered the CI on Darwin so when that passes and when we are happy with Regula Falsi, I can force merge

**Comment by Yurlungur** [comment]
I ran `compare_to_eospac` with caching disabled on an x86-volta node to see the difference between Regula Falsi and the old root finder. Regula Falsi is about 25% faster on CPU and negligibly faster on GPU, probably reflecting that the GPU is not flop bound on the test I ran. However, I did have errors with Steel where Regula Falsi was unable to bound the root. 

@jdolence could you maybe look at that and see if your root binding code could be improved to be a bit more robust?

**Comment by Yurlungur** [comment]
> I ran `compare_to_eospac` with caching disabled on an x86-volta node to see the difference between Regula Falsi and the old root finder. Regula Falsi is about 25% faster on CPU and negligibly faster on GPU, probably reflecting that the GPU is not flop bound on the test I ran. However, I did have errors with Steel where Regula Falsi was unable to bound the root.
> 
> @jdolence could you maybe look at that and see if your root binding code could be improved to be a bit more robust?

Actually I saw the same issue on dry air as well.

**Comment by jdolence** [comment]
> > I ran `compare_to_eospac` with caching disabled on an x86-volta node to see the difference between Regula Falsi and the old root finder. Regula Falsi is about 25% faster on CPU and negligibly faster on GPU, probably reflecting that the GPU is not flop bound on the test I ran. However, I did have errors with Steel where Regula Falsi was unable to bound the root.
> > @jdolence could you maybe look at that and see if your root binding code could be improved to be a bit more robust?
> 
> Actually I saw the same issue on dry air as well.

looking at it.  thanks for checking this.

File: test/test_pte.cpp
Comment (dholladay00) [inline test/test_pte.cpp:131]: should be `RView`, the p9 test caught this. I can fix and push, just not sure how you feel about others pushing to your branch w/o you knowing.

Comment (jdolence) [reply test/test_pte.cpp:131]: sure go for it.  looking at the thing @Yurlungur pointed out.  thanks.

**Comment by jhp-lanl** [comment]
Running the CI on gitlab...

File: singularity-eos/eos/eos_base.hpp
Comment (dholladay00) [inline singularity-eos/eos/eos_base.hpp:265]: Does adding a function to the base class mean that our using statements need to change as well for CRTP?

Comment (jhp-lanl) [reply singularity-eos/eos/eos_base.hpp:265]: I'd think so, but doesn't the PTE test use analytic EOS, so wouldn't an issue be exposed there?

Comment (jhp-lanl) [reply singularity-eos/eos/eos_base.hpp:265]: The PTE test uses analytic EOS that should require this, so I'd think it would throw an error if there was an issue

Comment (dholladay00) [reply singularity-eos/eos/eos_base.hpp:265]: oh interesting, yeah the CRTP on cppreference doesn't have the usings so I guess not.

**Comment by Yurlungur** [comment]
I merged #121 into this branch so I could compare performance. I also discovered an issue on GPUs. In particular, the interpolation for the cold curve can be a little off. This is important because it triggers whether or not root finding is performed. If your energy is below the cold curve there may be no root, and the root finder can fail.

I added a "soft" threshold around floating point precision that triggers extrapolation in case interpolation of the cold curve is insufficiently accurate.

**Comment by jdolence** [comment]
OK, is this good to go then?  Anything else need doing before merging?

**Comment by Yurlungur** [comment]
Update on performance: I am seeing between a 5x (CPU) to 10x (GPU) improvement on root finds from regula falsi. I will do a small amount more cleanup and then merge.

Thanks for all these enhancements, @jdolence I think this is a big boost in capability/performance across the board.

**Comment by Yurlungur** [comment]
Triggered CI on re-git. Will merge after it passes.

**Comment by Yurlungur** [comment]
We're good. Merging. 🥳

**Pull request merged**
